[Mono]: Make EventPipe MonoProfiler provider GC events async safe.

lateralusX opened the pull request:
EventPipe MonoProfiler provider includes most Mono profiler events, but events currently triggered during GC is not emitted since EventPipe is not async safe.

This PR add support to emit the MonoProfiler events triggered during GC (gc resize, gc moves, gc roots and heap dump object references) into a async safe temporary memory buffer that will be emitted into EventPipe once GC completes (and world has been restarted). This opens up the ability to do heapshots using EventPipe running on Mono into nettrace files.

Heapshots can be triggered in same way as on CoreCLR, creating an EventPipe session with GCHeapCollectKeyword keyword. Only one heapshot can be triggered at a time, so in order to trigger an additional heapshot, a new EventPipe session is setup (after the previous heapshot has completed). Heapshot events will be included in all session with configured keywords.

It is also possible to include all vtable/class references (including class name), into heap dump. This is enabled by a specific keyword and makes the dump self contained (no need to track type load etc). That way runtime can start up without any profiler support, and then it is possible to use custom tool or dotnet-trace to request a heap dump (including all referenced vtable/class references) and once that session is done, it is possible to do a new session and get a new separate heap dump into a new session. Having separate sessions in different files opens up the ability to diff between any heap dump over time.

lambdageek commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -3848,11 +4974,15 @@ mono_profiler_jit_chunk_destroyed (
 	if (!EventEnabledMonoProfilerJitChunkDestroyed())
 		return;
 
+	mono_profiler_fire_event_enter ();
```
Do you think it's worthwhile to bake the calls to `fire_event_enter` / `fire_event_exit` into the `FireEtw_XYZ` methods?  Seems errorprone to write them by hand.

lateralusX replied:
This is macros generated from manifest file shared with CoreCLR so if we do something we need to special case but we also use the same macros when replaying buffered events and then you shouldn't do enter/exit. Since this is a limited number of locations and I don't expect us to add that many new events into this provider, I believe we could write them by hand for now.

lambdageek reviewed (approved):
lgtm.  I'd like to see a bit more abstraction around the shared/exclusive lock (particularly the casting and bit manipulation could use some macros or helper functions).

Also consider making it automatic to take the shared fire event lock by making it part of the FireXYZ generated methods.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -3642,143 +3838,1061 @@ EventPipeEtwCallbackDotNETRuntimeStress (
+	uint8_t *buffer = NULL;
+
+	if (!current_block) {
+		current_block = mono_profiler_mem_block_alloc (req_size);
```
Jupp.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -3642,143 +3838,1061 @@ EventPipeEtwCallbackDotNETRuntimeStress (
-		NULL,
-		NULL);
+	// Make sure fire event calls will block and wait for GC completion.
+	ep_rt_spin_lock_aquire (&_ep_rt_mono_profiler_gc_state_lock);
+
+	// Mark gc state in progress bits, preventing new fire event requests.
+	do {
+		old_state = ep_rt_volatile_load_uint32_t (&_ep_rt_mono_profiler_gc_state);
+		count = (uint16_t)old_state;
+		EP_ASSERT ((old_state >> 16) == 0);
+		new_state = (uint32_t)(MONO_PROFILER_GC_IN_PROGRESS_STATE_VALUE << 16) | (uint32_t)(count);
+	} while (mono_profiler_atomic_cas_uint32_t (&_ep_rt_mono_profiler_gc_state, old_state, new_state) != old_state);
```
Yes, and it should be fairly uncommon case to hit the slow path (only when we have a GC that will do GC events).

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -3642,143 +3838,1061 @@ EventPipeEtwCallbackDotNETRuntimeStress (
+		} else {
+			ep_rt_thread_sleep (200);
+		}
+		count = (uint16_t)ep_rt_volatile_load_uint32_t (&_ep_rt_mono_profiler_gc_state);
```
I see if I can make it clearer.

lateralusX committed:
Make EventPipe MonoProfiler provider GC events async safe.

EventPipe MonoProfiler provider includes most Mono profiler events,
but events currently triggered during GC is not emitted since EventPipe
is not async safe.

This commit add support to emit the MonoProfiler events triggered during
GC (gc resize, gc moves, gc roots and heap dump object references) into
a async safe temporary memory buffer that will be emitted into EventPipe
once GC completes (and world has been restarted). This opens up the
ability to do heapshots using EventPipe running on Mono into nettrace
files.

Heapshots can be triggered in same way as on CoreCLR, creating an
EventPipe session with GCHeapCollectKeyword keyword. Only one heapshot
can be triggered at a time, so in order to trigger an additional
heapshot, a new EventPipe session is setup (after the previous heapshot
has completed). Heapshot events will be included in all session with
configured keywords.

It is also possible to include all vtable/class references
(including class name), into heap dump. This is enabled by a specific
keyword and makes the dump self contained (no need to track type load etc).
That way runtime can start up without any profiler support, and then it
is possible to use custom tool or dotnet-trace to request a heap dump
(including all referenced vtable/class references) and once that
session is done, it is possible to do a new session and get a new
separate heap dump into a new session. Having separate sessions in
different files opens up the ability to diff between any heap dump over
time.

lateralusX committed:
Fix build errors.

lateralusX committed:
Review feedback.

lateralusX commented:
@lambdageek Rebase and address feedback doing all bit manipulation in defines.

lateralusX committed:
Fix build errors.

lateralusX committed:
Add MonoProvider heapcollect filter keywords.

lateralusX committed:
Fix build error.

lateralusX committed:
Fix build error.

lateralusX committed:
Add support for session filter param value into DumpHeapStart event.

lateralusX committed:
Fix build error.

lateralusX commented:
Failures in runtime-staging is unrelated and appears on several other PR's as well.
